lib/disk/modules/VMWareSparseDisk.rb

```diff
@@ -180,4 +295,42 @@ def findFreeSector
     raise "Disk full." if @freeSector * blockSize > @capacity
     @freeSector
   end
+
+  def read_compressed_grain(grain_offset)
+    @vmwareSparseDisk_file.seek(grain_offset, IO::SEEK_SET)
+    buffer = @vmwareSparseDisk_file.read(SIZEOF_SPARSE_EXTENT_COMPRESSED_GRAIN_HEADER)
+    compressed_grain_header = SPARSE_EXTENT_COMPRESSED_GRAIN_HEADER.decode(buffer)
+    buffer = @vmwareSparseDisk_file.read(compressed_grain_header['size'])
+    Zlib::Inflate.inflate(buffer)
+  end
+
+  def read_grain_from_disk(gte, offset, length)
+    grain_location = gte * BYTES_PER_SECTOR
+
+    if @flags[:compressed]
+      buffer = read_compressed_grain(grain_location)
+      buffer[offset, length]
+    else
+      @vmwareSparseDisk_file.seek(grain_location + offset, IO::SEEK_SET)
+      @vmwareSparseDisk_file.read(length)
+    end
+  end
+
+  def read_grain(grain_number, offset, length, parent_offset)
+    gte = getGTE(grain_number)
+
+    #
+    # when GTE is 0 - reads with no parent: return 0s
+    #               - reads with    parent: read from parent
+    # when GTE is 1 - all reads: return 0s
+    # when GTE >  1 - all reads: from the sparse disk
+    #
```

Comment:
Can you order the comment in the same order as the if/else conditions...I was trying to compare the 2 and it was confusing

Either that or...

```ruby
    if gte > 1
      # all reads: from the sparse disk
      ...
```

Reply:
re-ordered comments to better match the code